### PR TITLE
[DTensor] Reject random ops on mismatched local tensor and mesh devices

### LINKED ISSUE #180088: [DTensor] RNG tracker does not advance state for CPU tensors on CUDA mesh, causing trunc_normal_ infinite loop
### 🐛 Describe the bug

Bug

  DTensor's RNG tracker does not advance CPU RNG state when a DTensor has a CUDA mesh but its local tensor is on CPU. This causes nn.init.trunc_normal_ to loop forever since its rejection sampling (introduced in #174997)
  depends on normal_() producing different values each call.
                                                                                                                                                                                                                                   
  Root cause                                                                                                                                                                                                                     

  The DTensor RNG tracker advances the CUDA RNG but the local tensor is on CPU, so values come from the stale CPU RNG:                                                                                                             
   
  CUDA RNG advances: True (1st), True (2nd)                                                                                                                                                                                        
  CPU  RNG advances: False (1st), False (2nd)                                                                                                                                                                                      
                                                                                                                                                                                                                                   
  How to get a CPU DTensor on a CUDA mesh                                                                                                                                                                                          
                                                                                                                                                                                                                                   
  This happens in FSDP seed checkpoint creation:                                                                                                                                                                                   
  ```                                                                                                                                                                                                                               
  with torch.device("meta"):
      model = model_config.build()                                                                                                                                                                                                 
  fully_shard(model, mesh=cuda_mesh)   # DTensors with CUDA mesh
  model.to_empty(device="cpu")          # local tensors now on CPU                                                                                                                                                                 
  model.init_weights()                  # trunc_normal_ → infinite loop       
```                                                                                                                                                     
                                                                                                                                                                                                                                   
  Repro                                                                                                                                                                                                                            
```                                                                                                                                                                                                         
  # torchrun --nproc_per_node=1 repro.py                                                                                                                                                                                         
  import torch, torch.distributed as dist                                                                                                                                                                                          
  from torch.distributed.device_mesh import init_device_mesh
  from torch.distributed.tensor import DTensor, Shard                                                                                                                                                                              
                                                                                                                                                                                                                                 
  dist.init_process_group("nccl")                                                                                                                                                                                                  
  torch.cuda.set_device(0)                                                                                                                                                                                                       
                                                                                                                                                                                                                                   
  mesh = init_device_mesh("cuda", (1,))                                                                                                                                                                                            
  with torch.device("meta"):
      t = torch.empty(4)                                                                                                                                                                                                           
  dt = DTensor.from_local(t, mesh, [Shard(0)])
  dt = torch.empty_like(dt, device="cpu")

  torch.manual_seed(42)
  cpu_before = torch.random.get_rng_state()
  cuda_before = torch.cuda.get_rng_state()
  dt.normal_(0, 1)
  cpu_after = torch.random.get_rng_state()
  cuda_after = torch.cuda.get_rng_state()

  print(f"CUDA RNG advances: {not torch.equal(cuda_before, cuda_after)}")
  print(f"CPU  RNG advances: {not torch.equal(cpu_before, cpu_after)}")

  a = dt.normal_(0, 1).to_local().clone()
  b = torch.empty_like(dt).normal_(0, 1).to_local()
  assert not torch.equal(a, b), f"BUG: RNG not advancing, a == b == {a}"

  dist.destroy_process_group()
```

### Versions

pytorch nightly

cc @ezyang @gchanan @kadeng @msaroufim @awgu @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @pragupta @dcci @aditvenk @xmfan

### PR DESCRIPTION
**Pull request by zupengwang**
A CPU-local DTensor on an accelerator mesh enters the accelerator RNG tracker, which restores CPU RNG state and can repeatedly generate the same samples. Reject mismatched non-meta local devices before constructing the tracker or executing the random operation. Preserve meta initialization and matching-device behavior.

Add regressions for random operators, placements, tracker initialization, input and RNG preservation, and the FSDP meta-to-CPU initialization path. This follows the error-on-mismatch approach proposed in closed PR #180087 and fixes #180088.

Test Plan
=========
On the remote RTX 3090 host:
```bash
bash /ch_data/wzp/oss-ai-infra/pytorch-dtensor-rng-device-20260919/run_gpu.sh fixed-verified test/distributed/tensor/test_random_ops.py -k device_mismatch -v

cd /ch_data/wzp/oss-ai-infra/pytorch-dtensor-rng-device-20260919/repo
env -u PYTHONPATH CUDA_VISIBLE_DEVICES= OMP_NUM_THREADS=2 ../.venv/bin/python test/distributed/tensor/test_random_ops.py DistTensorRandomInitTest.test_init_ops DistTensorRandomInitTestWithLocalTensor.test_init_ops -v

lintrunner -a
```
34 focused regressions and 2 CPU controls pass; 8 existing accelerator controls also pass. The added two-GPU regression fails on the unmodified source. Applicable lint checks pass. Validation used current-main Python source with existing CUDA 13.0 nightly native libraries, not a clean native build at this base. Full upstream CI and other accelerators were not run.

### CODE REVIEW

File: torch/distributed/tensor/_dispatch.py [hunk 1]
```diff
@@ -373,6 +373,22 @@ def _dispatch_get_local_results_slow_path(
             # run local op computation with potentially modified args/kwargs
             local_tensor_args = cast(tuple[object, ...], local_tensor_args)
             if op_call in self._random_ops:
+                first_arg, first_local_arg = (
+                    cast(dtensor.DTensor, args[0]),
+                    cast(torch.Tensor, local_tensor_args[0]),
+                )
+                if (
+                    not first_local_arg.is_meta
+                    and first_local_arg.device.type != mesh.device_type
```
Comment (ezyang) [inline torch/distributed/tensor/_dispatch.py:382]: [P2] Account for the output device of factory-like random ops

This checks the input device, but `rand_like` and `randn_like` can generate on an explicitly different output device. With the CPU-local DTensor on a CUDA mesh constructed in the new test, `torch.rand_like(dt, device="cuda")` previously entered the CUDA tracker's `_distribute_region` and generated CUDA output with the appropriate RNG. This guard now rejects it solely because the input is on CPU.

Please check the effective generation device for these factory-like operators, accounting for an explicit `device` override, and add regression coverage for CPU input with CUDA output (`rand_like` and `randn_like`).

Verified by tracing the source; I have not run a runtime reproduction.

— Astra agent
